[visionOS] Introduce a flag to delegate Image Fullscreen to Quick Look

etiennesegonzac opened the pull request:
#### be619ad087652b064ffa055bd66568035e58228a
<pre>
[visionOS] Introduce a flag to delegate Image Fullscreen to Quick Look
<a href="https://bugs.webkit.org/show_bug.cgi?id=268826">https://bugs.webkit.org/show_bug.cgi?id=268826</a>
&lt;<a href="https://rdar.apple.com/120903017">rdar://120903017</a>&gt;

Reviewed by Tim Horton.

On visionOS, when fullscreening an image, create a SharedMemory buffer of the cached
image and forward it to Quick Look.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
* Source/WebKit/Shared/FullScreenMediaDetails.h: Added.
* Source/WebKit/Shared/FullScreenMediaDetails.serialization.in: Added.
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.messages.in:
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundlePageFullScreenClient.cpp:
(WebKit::InjectedBundlePageFullScreenClient::enterFullScreenForElement):
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundlePageFullScreenClient.h:
Introduce a new serializable FullScreenMediaDetails struct that contains
all the media information needed for visionOS fullscreen.

* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
(WebKit::WebFullScreenManagerProxy::isVideoElement const):
(WebKit::WebFullScreenManagerProxy::isImageElement const):
(WebKit::WebFullScreenManagerProxy::imageMIMEType const):
(WebKit::WebFullScreenManagerProxy::imageBuffer const):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::enterFullScreen):
Create a SharedBuffer from the image SharedMemoryHandle.
(WebKit::WebFullScreenManagerProxy::exitFullScreen):
Release the SharedBuffer on exit.
(WebKit::WebFullScreenManagerProxy::isVideoElement const): Deleted.
Moved to the header with other similar methods.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/WebKitSwift.h:
* Source/WebKit/WebKitSwift/Preview/WKSPreviewWindowController.h: Added.
* Source/WebKit/WebKitSwift/Preview/PreviewWindowController.swift: Added.
(PreviewWindowController.item):
(PreviewWindowController.previewApp):
(PreviewWindowController.delegate):
(PreviewWindowController.presentWindow):
(PreviewWindowController.dismissWindow):
Add a Swift wrapper for the AssetViewer Preview SPI.

* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
Add WKSPreviewWindowController to WebKitSoftLink.
* Source/WebKit/SourcesCocoa.txt:
Mark the WebKitSoftLink with @no-unify.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
Soft link libWebKitSwift to access the new WKSPreviewWindowController.
(-[WKFullScreenWindowController _performSpatialFullScreenTransition:completionHandler:]):
(-[WKFullScreenWindowController showUI]):
Launch a Quick Look preview and keep the inWindow faded when Image data
is available.
Always hide the chrome when a Quick Look Preview is displayed.
(-[WKFullScreenWindowController previewWindowControllerDidClose]):
(-[WKFullScreenWindowController provideDataForItem:]):
Adopt the WKSPreviewWindowControllerDelegate and
QLPreviewItemDataProvider protocols.
Create an NSData object from the manager&apos;s SharedBuffer to provide the
image to Quick Look.

* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::enterFullScreenForElement):
When the new preference is on, and the fullscreen element is an image,
create a SharedMemory buffer of the cached image and forward it.

Canonical link: <a href="https://commits.webkit.org/274816@main">https://commits.webkit.org/274816@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/041d4b410f5b4784a482fcbb0baa83e33134cb92

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/39921 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/18932 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/42296 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/42465 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/35823 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/22025 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/16262 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/33358 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/40495 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/16092 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/34522 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/13811 "Passed tests") | 
| [✅ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/39789 "Passed tests") | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/13959 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/35477 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/43744 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/33374 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/36380 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/35815 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/39573 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/39547 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/14736 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/12128 "Passed tests") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/37959 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/16355 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/46555 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/8999 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/16404 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/9578 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/15999 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -61,6 +61,11 @@
+#import <pal/ios/QuickLookSoftLink.h>
+#import "WebKitSwiftSoftLink.h"
+SOFT_LINK_CLASS_FOR_HEADER(WebKit, WKSPreviewWindowController)
+SOFT_LINK_CLASS_FOR_SOURCE_OPTIONAL(WebKit, WebKitSwift, WKSPreviewWindowController)
```
This should go into WebKitSwiftSoftLink.h

aestes commented on Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
```diff
@@ -127,6 +135,8 @@ class WebFullScreenManagerProxy : public IPC::MessageReceiver {
 #if PLATFORM(VISION)
     bool m_isVideoElement { false };
+    String m_imageMimeType { };
+    std::optional<RetainPtr<NSData>> m_imageData { };
```
Nit: `{ }`s are unnecessary here.

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1856,16 +1874,37 @@ - (void)_performSpatialFullScreenTransition:(BOOL)enter completionHandler:(Compl
+    if (manager && manager->isImageElement()) {
+        if (enter) {
+            inWindowAlpha = 0;
+            NSString *appDisplayName = [[NSBundle mainBundle] objectForInfoDictionaryKey:(__bridge NSString *)_kCFBundleDisplayNameKey];
```
Should this use `applicationVisibleName()` here? That handles the edge case where a bundle doesn't have a display name.

aestes commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -61,6 +61,11 @@
 #import "MRUIKitSPI.h"
+#import "WKSPreviewWindowController.h"
+#import <pal/ios/QuickLookSoftLink.h>
+#import "WebKitSwiftSoftLink.h"
```
When we import a soft-linking header we need to mark the file with `@no-unify` in SourcesCocoa.txt and add it to the WebKit target in Xcode.

etiennesegonzac replied:
Should the header also be added to the WebKit target? (the impl already was)

aestes commented on Source/WebKit/WebKitSwift/Preview/PreviewWindowController.swift:
```diff
@@ -0,0 +1,71 @@
+@_spi(Preview) import AssetViewer
+
+@objc(WKSPreviewWindowController)
+public final class PreviewWindowController: NSObject {
```
FWIW the compiler can check that this implementation matches what's in WKSPreviewWindowController.h by making this an `@_objcImplementation extension WKSPreviewWindowController {`. Just curious if that works here (but if not it's fine as-is).

aestes replied:
Nevermind, that probably won't work because of `var previewApp: AssetViewer.Preview?` (unless that type is a class that derives from `NSObject`).

etiennesegonzac replied:
this might change too, making a note to try it once we have the final version!

pxlcoder commented on Source/WebKit/Shared/FullScreenMediaDetails.h:
```diff
@@ -0,0 +1,48 @@
+namespace IPC {
+class Decoder;
+class Encoder;
+}
```
These don't appear to be used here? Are they necessary?

etiennesegonzac replied:
my bad, I thought they were needed for the serialization to work.

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1856,16 +1874,37 @@ - (void)_performSpatialFullScreenTransition:(BOOL)enter completionHandler:(Compl
 
             scene.mrui_placement.preferredChromeOptions = [originalState sceneChromeOptions];
+            if (isImageElement)
+                scene.mrui_placement.preferredChromeOptions = RSSSceneChromeOptionsNone;
```
Maybe this should only be done in the `if (enter)` section above?

etiennesegonzac replied:
whoops, missed this comment but I'll get to it in the next round :)

pxlcoder commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1901,6 +1942,26 @@ - (void)hideUI
 @end
 
+#if PLATFORM(VISION)
+@implementation WKFullScreenWindowController (WKSPreviewWindowControllerDelegate)
```
Nit: Add a newline before this line.

pxlcoder commented on Source/WebKit/WebKitSwift/Preview/WKSPreviewWindowController.h:
```diff
@@ -0,0 +1,49 @@
+
+#import <Foundation/Foundation.h>
+
+#if defined(TARGET_OS_VISION) && TARGET_OS_VISION
```
`#if PLATFORM(VISION)` (unless there's a reason we can't use our macros in `WebKitSwift`?)

aestes replied:
We shouldn't use our macros in these headers.

hortont424 commented on Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
```diff
@@ -77,7 +82,10 @@ class WebFullScreenManagerProxy : public IPC::MessageReceiver {
+    bool isVideoElement() const { return m_isVideoElement; }
+    bool isImageElement() const { return m_imageData.has_value(); }
+    const String& imageMIMEType() const { return m_imageMIMEType; }
+    const std::optional<RetainPtr<NSData>>& imageData() const { return m_imageData; }
```
This Cocoa type feels quite out of place here.

hortont424 commented on Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
```diff
@@ -1858,17 +1875,37 @@ - (void)_performSpatialFullScreenTransition:(BOOL)enter completionHandler:(Compl
+            inWindowAlpha = 0;
+            auto item = adoptNS([PAL::allocQLItemInstance() initWithDataProvider:self contentType:[UTType typeWithMIMEType:manager->imageMIMEType()].identifier previewTitle:WebKit::applicationVisibleName()]);
+            _previewWindowController = [WebKit::allocWKSPreviewWindowControllerInstance() initWithItem:item.get()];
+            _previewWindowController.get().delegate = self;
```
square brackets over .get().

hortont424 reviewed (approved):
Does this whole thing need a compile-time flag?

etiennesegonzac committed:
[visionOS] Introduce a flag to delegate Image Fullscreen to Quick Look
https://bugs.webkit.org/show_bug.cgi?id=268826
<rdar://120903017>

Reviewed by Tim Horton.

On visionOS, when fullscreening an image, create a SharedMemory buffer of the cached
image and forward it to Quick Look.

* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/webkit/messages.py:
(types_that_must_be_moved):
* Source/WebKit/Shared/FullScreenMediaDetails.h: Added.
* Source/WebKit/Shared/FullScreenMediaDetails.serialization.in: Added.
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.messages.in:
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundlePageFullScreenClient.cpp:
(WebKit::InjectedBundlePageFullScreenClient::enterFullScreenForElement):
* Source/WebKit/WebProcess/InjectedBundle/InjectedBundlePageFullScreenClient.h:
Introduce a new serializable FullScreenMediaDetails struct that contains
all the media information needed for visionOS fullscreen.

* Source/WebKit/UIProcess/WebFullScreenManagerProxy.h:
(WebKit::WebFullScreenManagerProxy::isVideoElement const):
(WebKit::WebFullScreenManagerProxy::isImageElement const):
(WebKit::WebFullScreenManagerProxy::imageMIMEType const):
(WebKit::WebFullScreenManagerProxy::imageBuffer const):
* Source/WebKit/UIProcess/WebFullScreenManagerProxy.cpp:
(WebKit::WebFullScreenManagerProxy::enterFullScreen):
Create a SharedBuffer from the image SharedMemoryHandle.
(WebKit::WebFullScreenManagerProxy::exitFullScreen):
Release the SharedBuffer on exit.
(WebKit::WebFullScreenManagerProxy::isVideoElement const): Deleted.
Moved to the header with other similar methods.

* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Source/WebKit/WebKitSwift/WebKitSwift.h:
* Source/WebKit/WebKitSwift/Preview/WKSPreviewWindowController.h: Added.
* Source/WebKit/WebKitSwift/Preview/PreviewWindowController.swift: Added.
(PreviewWindowController.item):
(PreviewWindowController.previewApp):
(PreviewWindowController.delegate):
(PreviewWindowController.presentWindow):
(PreviewWindowController.dismissWindow):
Add a Swift wrapper for the AssetViewer Preview SPI.

* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.h:
* Source/WebKit/UIProcess/Cocoa/WebKitSwiftSoftLink.mm:
Add WKSPreviewWindowController to WebKitSoftLink.
* Source/WebKit/SourcesCocoa.txt:
Mark the WebKitSoftLink with @no-unify.

* Source/WebKit/UIProcess/ios/fullscreen/WKFullScreenWindowControllerIOS.mm:
Soft link libWebKitSwift to access the new WKSPreviewWindowController.
(-[WKFullScreenWindowController _performSpatialFullScreenTransition:completionHandler:]):
(-[WKFullScreenWindowController showUI]):
Launch a Quick Look preview and keep the inWindow faded when Image data
is available.
Always hide the chrome when a Quick Look Preview is displayed.
(-[WKFullScreenWindowController previewWindowControllerDidClose]):
(-[WKFullScreenWindowController provideDataForItem:]):
Adopt the WKSPreviewWindowControllerDelegate and
QLPreviewItemDataProvider protocols.
Create an NSData object from the manager's SharedBuffer to provide the
image to Quick Look.

* Source/WebKit/WebProcess/FullScreen/WebFullScreenManager.cpp:
(WebKit::WebFullScreenManager::enterFullScreenForElement):
When the new preference is on, and the fullscreen element is an image,
create a SharedMemory buffer of the cached image and forward it.

Canonical link: https://commits.webkit.org/274816@main
